src/OpenLoco/src/Ui/Widget.h

```diff
@@ -174,27 +203,90 @@ namespace OpenLoco::Ui
         return makeWidget({ xPos, yPos }, { width, height }, WidgetType::button, colour, StringIds::stepper_plus, tooltip);
     }
 
-    constexpr Widget makeTextWidget(Ui::Point origin, Ui::Size size, WidgetType type, WindowColour colour, StringId content, StringId tooltip = StringIds::null)
+    constexpr Widget makeTextWidget(Ui::Point32 origin, Ui::Size32 size, WidgetType type, WindowColour colour, StringId content, StringId tooltip = StringIds::null)
     {
-        Widget out{};
-        out.left = origin.x;
-        out.right = origin.x + size.width - 1;
-        out.top = origin.y;
-        out.bottom = origin.y + size.height - 1;
-        out.type = type;
-        out.windowColour = colour;
-        out.text = content;
-        out.tooltip = tooltip;
-
+        Widget out{ origin, size, type, colour, content, tooltip };
         return out;
     }
 
-    constexpr Widget widgetEnd()
+    namespace Detail
     {
-        Widget out{};
-        out.type = WidgetType::end;
+        template<typename T, typename Enable = void>
+        struct WidgetsCount
+        {
+            static constexpr size_t count = 0;
+        };
 
-        return out;
+        template<typename T>
+        struct WidgetsCount<T, std::enable_if_t<std::is_base_of_v<Widget, T>>>
+        {
+            static constexpr size_t count = 1;
+        };
+
+        template<typename T, std::size_t N>
+        struct WidgetsCount<std::array<T, N>, std::enable_if_t<std::is_base_of_v<Widget, T>>>
+        {
+            static constexpr size_t count = N;
+        };
+
+        template<typename T>
+        struct IsWidgetsArray : std::false_type
+        {
+        };
+
+        template<std::size_t N>
+        struct IsWidgetsArray<std::array<Widget, N>> : std::true_type
+        {
+        };
     }
 
+    template<typename... TArgs>
+    constexpr auto makeWidgets(TArgs&&... args)
+    {
+        constexpr auto totalCount = [&]() {
+            size_t count = 0;
+            ((count += Detail::WidgetsCount<std::decay_t<decltype(args)>>::count), ...);
+            return count;
+        }();
+
+        std::array<Widget, totalCount> res{};
+        size_t index = 0;
+
+        const auto append = [&](auto&& val) {
+            if constexpr (Detail::IsWidgetsArray<std::decay_t<decltype(val)>>::value)
+            {
+                for (auto&& widget : val)
+                {
+                    res[index] = std::move(widget);
+                    index++;
+                }
+            }
+            else
+            {
+                res[index] = std::move(val);
+                index++;
+            }
+        };
+
+        ((append(args)), ...);
+
+        return res;
+    }
+
+    constexpr auto makeStepperWidget(Ui::Point32 origin, Ui::Size32 size, WindowColour colour, uint32_t content = Widget::kContentNull, StringId tooltip = StringIds::null)
```

Comment:
I think we should keep this plural (Widget_s_) while it still creates three widgets, just so there's no misunderstanding on the widget index front.

Reply:
I guess you are right, the long term goal is to have a single widget for this purpose but you are correct that its technically doing more than one.